### DISCUSSION TITLE
Subscription event context

### DISCUSSION THREAD

**Discussion by BunkenU** [graphql-java/graphql-java#3712] [Q&A]
Hello!
My use case is that I need to attach some extra information to each individual subscription event produced by the publisher in my data fetcher. This is to perform some tasks once the event has been successfully mapped to an execution result and is ready to be delivered to the client. 

The publisher comes from another part of the application as a `Flux`, so my idea was to transform emitted elements by wrapping them in a `DataFetcherResult`, where I put exactly what I need in the local context. And then use an `Instrumentation` with the `beginSubscribedFieldEvent` method, and perform my tasks in the `onComplete` callback. I could of course do it once the event has left the engine as well but at that point I only have the final `ExecutionResult` (?).

It seems we don't have access to local context in the instrumentation method. I see in `SubscriptionExecutionStrategy` that a new `ExecutionContext` is prepared for each event, however it's the overall execution context that is passed to the parameters. But that alone won't be enough since `unboxPossibleDataFetcherResult` happens after the instrumentation is called (though I could do it myself if I have the root eventPayload).

Am I missing something and/or does anyone have any other ideas how to accomplish this?

**Comment by bbakerman** [comment]
> My use case is that I need to attach some extra information to each individual subscription event produced by the publisher in my data fetcher. 

What extra data do your want to attach?  Is this to be expressed as graphql data?

eg it starts as `record Foo(String name, int age)` and you want to attach more data so its logically `record Bar(String name, int age, int height)`. ??

if so why not use `flux.map(foo -> Bar.from(foo, andHeight))` ?

eg use Flux mapping not graphql interception?  You can "instrument the DataFecther" such that you will always see "flux values" coming from the DataFetcher

`graphql.execution.instrumentation.Instrumentation#instrumentDataFetcher` say

**Reply by BunkenU** [reply]
Thanks for the answer! 
The extra data is not to be expressed as graphql data. Basically I need to audit log (and some other things) information about each object being delivered over the long lived subscription as it happens.

I have a central component that handles the transport, eg consumes the `SubscriptionPublisher` returned from `graphql.execute().getData()` by delivering each `ExecutionResult` to the client. 
And then multiple data fetchers for different types of objects, but all finally handled at the central component.

So for a `record Foo(int id, String name, int age)` the grahpql query may only select name but I need id and type=Foo etc. 

There is no problem preparing and logging this in the data fetchers or by instrumenting data fetcher where i can access the Foo, but I feel like that is too early? Since child fields etc has not been resolved, ideally I'd like to access or compute the info once the overall execution result has been completed successfully

**Reply by bbakerman** [reply]
I think you plan is in trouble - a completed "subscription field" is  either a `CompletableFuture<Map<String, Object>> | Map<String, Object>`. so basically a map or a promise to a map (if its async)

So `graphql.execution.instrumentation.Instrumentation#beginFieldCompletion` returns when all the sub fields are done but you only have a map.  You could do it then for subscriptions.  Every top level subscription field MUST be an object type and hence a map

This then gets put into a ExecutionResult as each "event" is sent back down the publisher.  

But you can change the object and do dont get the original "source" object the original fetcher produced say.
